odml/templates.py

```diff
@@ -0,0 +1,188 @@
+"""
+Handles (deferred) loading of odML templates
+"""
+
+import os
+import sys
+import tempfile
+import threading
+try:
+    import urllib.request as urllib2
+    from urllib.error import URLError
+    from urllib.parse import urljoin
+
+except ImportError:
+    import urllib2
+    from urllib2 import URLError
+    from urlparse import urljoin
+
+from datetime import datetime as dati
+from datetime import timedelta
+from hashlib import md5
+
+from .tools.parser_utils import ParserException
+from .tools.xmlparser import XMLReader
+
+
+REPOSITORY_BASE = 'https://templates.g-node.org/'
+REPOSITORY = urljoin(REPOSITORY_BASE, 'templates.xml')
+
+CACHE_AGE = timedelta(days=1)
+CACHE_DIR = "odml.cache"
+
+
+# TODO after prototyping move functions common with
+# terminologies to a common file.
+
+
+def cache_load(url):
+    """
+    Load the url and store the file in a temporary cache directory.
+    Subsequent requests for this url will use the cached version until
+    the file is older than the CACHE_AGE.
+
+    Exceptions are caught and not re-raised to enable loading of nested
+    odML files without breaking if one of the child files is unavailable.
+
+    :param url: location of an odML template XML file.
+    :return: Local file location of the requested file.
+    """
+
+    filename = '.'.join([md5(url.encode()).hexdigest(), os.path.basename(url)])
+    cache_dir = os.path.join(tempfile.gettempdir(), CACHE_DIR)
+
+    # Create temporary folder if required
+    if not os.path.exists(cache_dir):
+        try:
+            os.makedirs(cache_dir)
+        except OSError:  # might happen due to concurrency
+            if not os.path.exists(cache_dir):
+                raise
+
+    cache_file = os.path.join(cache_dir, filename)
+
+    if not os.path.exists(cache_file) or dati.fromtimestamp(os.path.getmtime(cache_file)) < (dati.now() - CACHE_AGE):
+        try:
+            data = urllib2.urlopen(url).read()
+            if sys.version_info.major > 2:
+                data = data.decode("utf-8")
+        except (ValueError, URLError) as exc:
+            print("Failed to load resource from '%s': %s" % (url, exc))
+            raise
+
+        with open(cache_file, "w") as local_file:
+            local_file.write(str(data))
+
+    return cache_file
+
+
+class TemplateHandler(dict):
```

Comment:
Subclassing the `dict` built-in is generally discouraged (and `collections.UserDict` exists for this purpose) but I think the reasons for that don't affect us here right now.  I believe it's only an issue when overriding the magic methods, which we're not.  Might want to keep in mind in case we do anything more fancy with the class.